lms/djangoapps/completion/tests/test_models.py

```diff
@@ -142,3 +144,146 @@ def test_cannot_call_submit_completion(self):
                 completion=0.9,
             )
         self.assertEqual(models.BlockCompletion.objects.count(), 1)
+
+
+@ddt.ddt
+class AggregateCompletionTestCase(TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super(AggregateCompletionTestCase, cls).setUpClass()
+        cls.overrider = waffle.waffle().override(waffle.ENABLE_COMPLETION_TRACKING, True)
+        cls.overrider.__enter__()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.overrider.__exit__(None, None, None)
+        super(AggregateCompletionTestCase, cls).tearDownClass()
+
+    def setUp(self):
+        super(AggregateCompletionTestCase, self).setUp()
+        self.user = UserFactory()
+
+    @ddt.data(
+        # Valid arguments
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', 0.5, 1, 0.5),
+    )
+    @ddt.unpack
+    def test_submit_completion_with_valid_data(self, block_key, aggregate_name, earned, possible, expected_percent):
+        block_key_obj = UsageKey.from_string(block_key)
+        obj, is_new = models.AggregateCompletion.objects.submit_completion(
+            self.user,
+            block_key_obj.course_key,
+            block_key_obj,
+            aggregate_name,
+            earned,
+            possible
+        )
+        self.assertTrue(is_new)
+        self.assertEqual(len(models.AggregateCompletion.objects.all()), 1)
+        self.assertEqual(obj.earned, earned)
+        self.assertEqual(obj.possible, possible)
+        self.assertEqual(obj.percent, expected_percent)
+
+    @ddt.data(
+        # Earned greater than possible
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', 1.1, 1,
+         ValueError, "The earned cannot be larger than the possible value."),
+        # Earned is less than zero.
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', -0.5, 1,
+         ValidationError, "{'percent': [u'-0.5 must be between 0.0 and 1.0'],"
+                          " 'earned': [u'-0.5 must be larger than 0.']}"),
+        # Possible is less than zero.
+        (u'block-v1:edx+test+run+type@video+block@doggos', 'course', -1.5, -1,
+         ValidationError, "{'percent': [u'1.5 must be between 0.0 and 1.0'],"
+                          " 'possible': [u'-1.0 must be larger than 0.'],"
+                          " 'earned': [u'-1.5 must be larger than 0.']}"),
+    )
```

Comment:
Please add tests to cover all the other error cases, including the `TypeError`s and `RuntimeError`.

tip: you'll need to test the waffle `RuntimeError` in a separate method, and run `cls.overrider = waffle.waffle().override(waffle.ENABLE_COMPLETION_TRACKING, False)` to un-set the switch.

Reply:
@macornwell  Thanks for adding those `TypeError` tests, but I had to add the `RuntimeError` test case with  94153c7601.